pkg/service/association.go

```diff
@@ -10,6 +12,46 @@ import (
 	"github.com/free5gc/smf/internal/pfcp/message"
 )
 
+func toBeAssociatedWithUPF(ctx context.Context, upf *smf_context.UPF) {
+	var upfStr string
+	if upf.NodeID.NodeIdType == pfcpType.NodeIdTypeFqdn {
+		upfStr = fmt.Sprintf("[%s](%s)", upf.NodeID.FQDN, upf.NodeID.ResolveNodeIdToIp().String())
+	} else {
+		upfStr = fmt.Sprintf("[%s]", upf.NodeID.ResolveNodeIdToIp().String())
+	}
+	ensureSetupPfcpAssociation(ctx, upf, upfStr)
+}
+
+func isDone(ctx context.Context) bool {
+	select {
+	case <-ctx.Done():
+		return true
+	default:
+		return false
+	}
+}
+
+func ensureSetupPfcpAssociation(ctx context.Context, upf *smf_context.UPF, upfStr string) {
+	var alertTime time.Time
+	for {
+		alertInterval := smf_context.SMF_Self().AssociationSetupFailedAlertInterval
+		err := setupPfcpAssociation(upf, upfStr)
+		if err == nil {
+			return
+		}
+		now := time.Now()
+		if alertTime.IsZero() || now.After(alertTime.Add(alertInterval)) {
+			logger.AppLog.Errorf("Failed to setup an association with UPF%s, error:%+v", upfStr, err)
+			alertTime = now
+		}
+
+		if isDone(ctx) {
+			logger.AppLog.Infof("Canceled association request to UPF%s", upfStr)
+			return
+		}
+	}
```

Comment:
This is a busy loop and can add Sleep 1 second here.

Reply:
I don't think so. Because PFCP Association Setup Request is retransmitted at the interval defined as `ResendRequestTimeOutPeriod`(= 3s) in pfcp library.

Reply:
PFCP retransmission timeout 1s and max retry 3 times by default.
Did you mean Sleep 1s too long? How about 500ms? Busy loop issue need to be resolved.

Reply:
but this routine will be blocked in `now.After(alertTime.Add(alertInterval)`. It will cause `ctx.Done()` can't be handled immediately.

Reply:
```
for {
	err := setupPfcpAssociation(upf, upfStr)
	if err == nil {
		return
	}

	alertInterval := smf_context.SMF_Self().AssociationSetupFailedAlertInterval
	now := time.Now()
	select {
	case <-now.After(alertTime.Add(alertInterval)):
		logger.AppLog.Errorf("Failed to setup an association with UPF%s, error:%+v", upfStr, err)
	case <-ctx.Done():
		logger.AppLog.Infof("Canceled association request to UPF%s", upfStr)
		return
	}
}
```

Reply:
`now.After(...)` is `func After` of `type Time`(see https://pkg.go.dev/time#Time.After) , not `func After` of `package time`.(see https://pkg.go.dev/time#After).
So this routine is not blocked.

Reply:
got it